DigitalLearningSolutions.Data/Services/NotificationService.cs

```diff
@@ -1,61 +1,74 @@
-                                    <p>Dear {unlockData?.ContactForename}</p>
-                                    <p>Digital Learning Solutions Delegate, {unlockData?.DelegateName}, has requested that you unlock their progress for the course {unlockData?.CourseName}</p>
-                                    <p>They have reached the maximum number of assessment attempt allowed without passing.</p><p>To review and unlock their progress, <a href='{unlockUrl.Uri}'>click here</a>.</p>
-                                </body>"
-            };
-
-            emailService.SendEmail(new Email(emailSubjectLine, builder, unlockData.ContactEmail, unlockData.DelegateEmail));
-        }
-    }
-}
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System;
+    using DigitalLearningSolutions.Data.Models;
+    using DigitalLearningSolutions.Data.Models.Email;
+    using Microsoft.FeatureManagement;
+    using MimeKit;
+
+    public interface INotificationService
+    {
+        void SendUnlockRequest(int progressId);
+    }
+
+    public class NotificationService : INotificationService
+    {
+        private readonly IConfigService configService;
+        private readonly IEmailService emailService;
+        private readonly IFeatureManager featureManager;
+        private readonly INotificationDataService notificationDataService;
+
+        public NotificationService(
+            INotificationDataService notificationDataService,
+            IConfigService configService,
+            IEmailService emailService,
+            IFeatureManager featureManager
+        )
+        {
+            this.notificationDataService = notificationDataService;
+            this.configService = configService;
+            this.emailService = emailService;
+            this.featureManager = featureManager;
+        }
+
+        public void SendUnlockRequest(int progressId)
+        {
+            var unlockData = notificationDataService.GetUnlockData(progressId);
+            if (unlockData == null)
+            {
+                throw new NotificationDataException(
+                    $"No record found when trying to fetch Unlock Data. Progress ID: {progressId}"
+                );
+            }
+
+            unlockData.ContactForename = unlockData.ContactForename == "" ? "Colleague" : unlockData.ContactForename;
+            var refactoredTrackingSystemEnabled = featureManager.IsEnabledAsync("RefactoredTrackingSystem").Result;
+            var baseUrlConfigOption = refactoredTrackingSystemEnabled ? ConfigService.AppBaseUrl : ConfigService.TrackingSystemBaseUrl;
+            var trackingSystemBaseUrl = configService.GetConfigValue(baseUrlConfigOption) ??
+                                        throw new ConfigValueMissingException(
+                                            configService.GetConfigValueMissingExceptionMessage(baseUrlConfigOption)
+                                        );

```

Comment:
I don't understand why we are using this over Configuration.GetCurrentSystemBaseUrl and Configuration.GetAppRootPath in ConfigHelper which pulls the base url out of the appSetting.json

This will mean that we have to build the url more carefully, as the values won't have the /tracking/ part in them when we get them from the database. That this is currently incorrect with feature flag on based on the migration anyway, it'll generate https://hee-dls-test.softwire.com/coursedelegate?CustomisationID=xxxx instead of e.g. https://hee-dls-test.softwire.com/TrackingSystem/Delegates/CourseDelegates?CustomisationId=19262

Reply:
Should I get rid of the migration and just use the CurrentSystemBaseUrl from appsettings?

Reply:
CurrentSystemBaseUrl is the old site base url, we could keep that one as coming from the database, since it is an existing value and would mean we don't have to change the url calculation quite as much, but the new site url should not come from the migration, use the AppRootPath from config.

So yes, remove the migration (I guess manually delete the record on the test mbdbx101_test db since the migration will have run with the jenkins build).